internal/worker/worker.go

```diff
@@ -0,0 +1,444 @@
+	errInventoryQuery = errors.New("inventory query returned error")
+
+	errCollector = errors.New("collector error")
+)
+
+type Worker struct {
+	repository   store.Repository
+	stream       events.Stream
+	id           registry.ControllerID
+	cfg          *app.Configuration
+	syncWG       *sync.WaitGroup
+	logger       *logrus.Logger
+	name         string
+	facilityCode string
+	replicaCount int
+	concurrency  int
+	dispatched   int32
+}
+
+func New(
+	ctx context.Context,
+	facilityCode string,
+	stream events.Stream,
+	cfg *app.Configuration,
+	syncWG *sync.WaitGroup,
+	logger *logrus.Logger,
+) (*Worker, error) {
+	id, _ := os.Hostname()
+
+	concurrency := concurrency
+	if cfg.Concurrency != 0 {
+		concurrency = cfg.Concurrency
+	}
+
+	repository, err := store.NewRepository(ctx, cfg.StoreKind, model.AppKindOutOfBand, cfg, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Worker{
+		name:         id,
+		facilityCode: facilityCode,
+		cfg:          cfg,
+		syncWG:       syncWG,
+		logger:       logger,
+		repository:   repository,
+		stream:       stream,
+		concurrency:  concurrency,
+	}, nil
+}
+
+func (w *Worker) Run(ctx context.Context) {
+	tickerFetchEvents := time.NewTicker(fetchEventsInterval).C
+
+	if err := w.stream.Open(); err != nil {
+		w.logger.WithError(err).Error("event stream connection error")
+		return
+	}
+
+	// returned channel ignored, since this is a Pull based subscription.
+	_, err := w.stream.Subscribe(ctx)
+	if err != nil {
+		w.logger.WithError(err).Error("event stream subscription error")
+		return
+	}
+
+	w.logger.Info("connected to event stream.")
+
+	// register worker in NATS active-controllers kv bucket
+	w.startWorkerLivenessCheckin(ctx)
+
+	if _, err := createOrBindKVBucketWithOpts(w.stream, w.cfg.NatsOptions.KV); err != nil {
+		w.logger.WithError(err).Error("failed to create/bind to status kv" + inventoryStatusKVBucket)
+	}
+
+	v := version.Current()
+	w.logger.WithFields(
+		logrus.Fields{
+			"version":     v.AppVersion,
+			"commit":      v.GitCommit,
+			"branch":      v.GitBranch,
+			"concurrency": w.concurrency,
+		},
+	).Info("Alloy controller running")
+
+Loop:
+	for {
+		select {
+		case <-tickerFetchEvents:
+			if w.concurrencyLimit() {
+				continue
+			}
+
+			w.processEvents(ctx)
+
+		case <-ctx.Done():
+			if w.dispatched > 0 {
+				continue
+			}
+
+			break Loop
+		}
+	}
+}
+
+func (w *Worker) processEvents(ctx context.Context) {
+	// XXX: consider having a separate context for message retrieval
+	msgs, err := w.stream.PullMsg(ctx, 1)
+
+	switch {
+	case err == nil:
+	case errors.Is(err, nats.ErrTimeout):
+		w.logger.WithFields(
+			logrus.Fields{"err": err.Error()},
+		).Trace("no new events")
+	default:
+		w.logger.WithFields(
+			logrus.Fields{"err": err.Error()},
+		).Warn("retrieving new messages")
+
+		metrics.NATSError("pull-msg")
+	}
+
+	for _, msg := range msgs {
+		if ctx.Err() != nil || w.concurrencyLimit() {
+			w.eventNak(msg)
+
+			return
```

Comment:
If we return here, we leave messages in a delivered but not nak'd state. This becomes a bug when someone changes `PullMsg` to take a non `1` parameter (@ 160).

_Edit:_ Maybe the fix here is to change the semantics of `PullMsg` or provide a `PullSingleMsg` method? 🤔

Reply:
In the case they aren't nak'd, we fall back to the Jetstream consumer semantics to redeliver the message after the `ackWait` period, which is set to 5minutes - https://github.com/metal-toolbox/sandbox/blob/75846411ad772fd8520ab6a58cf4f17dcdffc713/templates/alloy-configmap.yaml#L21

Side note: I'll change the value to PullMsg to be the `concurrency` configured, since we're only polling the JS at 10 second intervals and Alloy can take on a lot more collection work in parallel.

Another side note re `PullMsg`: I think we need a method like `PullMsgFromConsumer("consumer-name", "configured.subject.string.")` which is more explicit and clear, and then deprecate `PullMsg`.